File: test_15_16/authenticator.java
import java.util.*;
import java.util.concurrent.*;

class Authenticator {
    public boolean authenticate(String user, String pwd) {
        return "admin".equals(user) && "secret".equals(pwd);
    }
}
class Encrypter {
    public String encrypt(String data) {
        StringBuilder sb = new StringBuilder();
        for (char c : data.toCharArray()) sb.append((char)(c + 2));
        return sb.toString();
    }
}
class Validator {
    public boolean validate(String input) {
        return input.matches("\\w+");
    }
}
class Logger {
    public void log(String msg) {
        System.out.println("SECLOG: " + msg);
    }
}

class AuthWorker implements Runnable {
    String username, password;
    Logger lg = new Logger();
    Authenticator auth = new Authenticator();
    Encrypter enc = new Encrypter();
    Validator val = new Validator();

    public AuthWorker(String user, String pwd) {
        username = user; password = pwd;
    }

    public void run() {
        if (!val.validate(username)) {
            lg.log("Validation failed for user");
            return;
        }
        if (auth.authenticate(username, password)) {
            String token = enc.encrypt(username + ":" + password);
            lg.log("Authenticated, token: " + token);
        } else {
            lg.log("Authentication failure");
        }
    }
}

public class MainApp {
    public static void main(String[] args) throws InterruptedException {
        List<String[]> users = Arrays.asList(
            new String[]{"admin", "secret"},
            new String[]{"guest", "test123"},
            new String[]{"super", "bob!234"}
        );
        ExecutorService pool = Executors.newFixedThreadPool(12);
        for (int i = 0; i < 25; i++) {
            for (String[] usr : users) {
                pool.submit(new AuthWorker(usr[0], usr[1]));
            }
            Thread.sleep(70);
        }
        pool.shutdown();
        pool.awaitTermination(3, TimeUnit.MINUTES);
    }
}

File: test_15_16/authenticator.py
import threading
import time
import random

class Authenticator:
    def authenticate(self, username, password):
        if username == "admin" and password == "secret":
            return True
        return False

class Encrypter:
    def encrypt(self, data):
        return ''.join(chr(ord(c) + 1) for c in data)

class Validator:
    def validate(self, data):
        return data.isalnum()

class Logger:
    def log(self, msg):
        print(f"LOG: {msg}")

class SecurityThread(threading.Thread):
    def __init__(self, user, pwd):
        threading.Thread.__init__(self)
        self.user = user
        self.pwd = pwd
        self.lg = Logger()
        self.auth = Authenticator()
        self.enc = Encrypter()
        self.val = Validator()

    def run(self):
        if not self.val.validate(self.user) or not self.val.validate(self.pwd):
            self.lg.log("Validation failed")
            return
        if self.auth.authenticate(self.user, self.pwd):
            token = self.enc.encrypt(f"{self.user}-{self.pwd}")
            self.lg.log(f"Authenticated, token: {token}")
        else:
            self.lg.log("Authentication failed")

def multi_thread_login(users_pwds):
    threads = []
    for user, pwd in users_pwds:
        t = SecurityThread(user, pwd)
        threads.append(t)
        t.start()
    for t in threads:
        t.join()

def main():
    users_pwds = [("admin", "secret"), ("user1", "pass1"), ("user2", "invalid!")]
    for _ in range(20):
        multi_thread_login(users_pwds)
        time.sleep(0.1)

if __name__ == "__main__":
    main()
